File: Searching/py/Binary Search.py
def Binary_Search(arr, first, last, x):

    if(last>=first):
        mid = (first + last)//2
        if (arr[mid] == x):
            return mid
        elif (arr[mid] > x):
            return Binary_Search(arr, first, mid - 1, x)
        else:
            return Binary_Search(arr, mid + 1, last, x)    

    else:
        return -1




arr1 = [10, 20, 30, 40]
x = int(input())

results = Binary_Search(arr1, 0, len(arr1)-1, x)
if results == -1:
    print("Not Found")
else:
    print("Number at", results+1)

File: Searching/cpp/binarySearch.cpp
#include <bits/stdc++.h>

using namespace std;

#define loop(i, a, b) for(int i = a; i < b; i++)
#define autoLoop(a, n) for(auto& a : n)
#define sze(x) (int)(x).size()

typedef long long ll;
typedef pair<int, int> pii;
typedef vector<int> vi;

void insertArray(vector<int> &nums) {
    int size, num;
    cin >> size;
    for (int i=0; i<size; i++) {
        cin >> num;
        nums.push_back(num);
    }
}

int binarySearch(vector<int> &nums, int first, int last, int findNum) {
    if (last >= first) {
        int mid = (first + last)/2;
        if (findNum == nums[mid])
            return mid;
        else if (findNum < nums[mid]) {
            return binarySearch(nums, 0, mid-1, findNum);
        }
        else if (findNum > nums[mid]) {
            return binarySearch(nums, mid+1, nums.size(), findNum);
        }
    }
    else {
        return -1;
    }
}

main() {
    vector <int> nums;
    int findNum;
    insertArray(nums);
    cin >> findNum;
    int first = 0, last = nums.size()-1;
    int ret = binarySearch(nums, first, last, findNum);
    if (ret == -1) 
        cout << "not found";
    else
        cout << "element found at " << ret << " position";
}